File: clojure/anagram/anagram.clj
(ns anagram)

(defn- sorted-lowered [s]
  (apply str (sort (clojure.string/lower-case s))))

(defn- is-anagram [s w]
  (and
    (not= s w)
    (= (sorted-lowered s) (sorted-lowered w))))

(defn anagrams-for [s words]
  (filter (partial is-anagram s) words))

File: haskell/anagram/anagram.hs
module Anagram (anagramsFor) where

import Data.List (sort)
import Data.Char (toLower)

normalise :: String -> String
normalise = sort . map toLower

anagramsFor :: String -> [String] -> [String]
anagramsFor s xs = filter go xs
  where
    go x = normalise s == normalise x && s /= x
